typescript/sdk/src/core/HyperlaneCore.ts

```diff
@@ -175,19 +165,129 @@ export class HyperlaneCore extends HyperlaneApp<CoreFactories> {
+      message: DispatchedMessage,
+      event: DispatchEvent,
+    ) => Promise<void>,
+    chains = Object.keys(this.contractsMap),
+  ): {
+    removeHandler: (chains?: ChainName[]) => void;
+  } {
+    chains.map((originChain) => {
+      const mailbox = this.contractsMap[originChain].mailbox;
+      this.logger.debug(`Listening for dispatch on ${originChain}`);
+      mailbox.on<DispatchEvent>(
+        mailbox.filters.Dispatch(),
+        (_sender, _destination, _recipient, message, event) => {
+          const parsed = HyperlaneCore.parseDispatchedMessage(message);
+          this.logger.info(`Observed message ${parsed.id} on ${originChain}`);
+          return handler(parsed, event);
+        },
+      );
+    });
+
+    return {
+      removeHandler: (removeChains) =>
+        (removeChains ?? chains).map((originChain) => {
+          this.contractsMap[originChain].mailbox.removeAllListeners('Dispatch');
+          this.logger.debug(`Stopped listening for dispatch on ${originChain}`);
+        }),
+    };
+  }
+
+  getDefaults(): Promise<ChainMap<{ ism: Address; hook: Address }>> {
+    return promiseObjAll(
+      objMap(this.contractsMap, async (_, contracts) => ({
+        ism: await contracts.mailbox.defaultIsm(),
+        hook: await contracts.mailbox.defaultHook(),
+      })),
+    );
+  }
+
+  getIsm(
+    destinationChain: ChainName,
+    recipientAddress: Address,
+  ): Promise<Address> {
+    const destinationMailbox = this.contractsMap[destinationChain];
+    return destinationMailbox.mailbox.recipientIsm(recipientAddress);
+  }
+
+  protected getRecipient(message: DispatchedMessage): IMessageRecipient {
+    return IMessageRecipient__factory.connect(
+      bytes32ToAddress(message.parsed.recipient),
+      this.multiProvider.getProvider(this.getDestination(message)),
+    );
+  }
+
+  estimateHandle(message: DispatchedMessage): Promise<ethers.BigNumber> {
+    return this.getRecipient(message).estimateGas.handle(
+      message.parsed.origin,
+      message.parsed.sender,
+      message.parsed.body,
+      { from: this.getAddresses(this.getDestination(message)).mailbox },
+    );
+  }
+
+  deliver(
     message: DispatchedMessage,
-    dispatchTx: ethers.ContractReceipt,
+    ismMetadata: string,
   ): Promise<ethers.ContractReceipt> {
-    const metadata = await this.buildMetadata(message, dispatchTx);
+    const destinationChain = this.getDestination(message);
+    return this.multiProvider.handleTx(
+      destinationChain,
+      this.getContracts(destinationChain).mailbox.process(
+        ismMetadata,
+        message.message,
+      ),
+    );
+  }
+
+  async getHook(
+    originChain: ChainName,
+    senderAddress: Address,
+  ): Promise<Address> {
+    const provider = this.multiProvider.getProvider(originChain);
+    try {
+      const client = MailboxClient__factory.connect(senderAddress, provider);
+      const hook = await client.hook();
+      if (!eqAddress(hook, ethers.constants.AddressZero)) {
```

Comment:
the `isZeroishAddress` util fn may be a more robust alternative for what you're trying to do here